File: cpp/MNN/test/test_selfAT_wh.cpp
#include "MSelfAT.hpp"
#include "test_common.hpp"
#include <iomanip>

using namespace mnn;
using namespace mtb;

#define FP_T float

void test_SelfAT(){
    START_TEST();
    {
        MSelfAT<FP_T> att(768, 16, 768/16);
        auto k_w = load_data<FP_T>("build/test/temp/k_proj_weight.bin", 768*768);
        auto v_w = load_data<FP_T>("build/test/temp/v_proj_weight.bin", 768*768);
        auto q_w = load_data<FP_T>("build/test/temp/q_proj_weight.bin", 768*768);
        auto o_w = load_data<FP_T>("build/test/temp/out_proj_weight.bin", 768*768);
        auto o_b = load_data<FP_T>("build/test/temp/out_proj_bias.bin", 768);
     
        auto it_s = load_data<FP_T>("build/test/temp/input_single.bin", 1*1*768);
        auto ot_s = load_data<FP_T>("build/test/temp/output_single.bin", 1*1*768);
        auto k_h = load_data<FP_T>("build/test/temp/k_h.bin", 1*16*5*48);
        auto v_h = load_data<FP_T>("build/test/temp/v_h.bin", 1*16*5*48);
        
        att.fill_k(k_w.data(), k_w.size());
        att.fill_v(v_w.data(), v_w.size());
        att.fill_q(q_w.data(), q_w.size());
        att.fill_out(o_w.data(), o_b.data(), o_w.size(), o_b.size());

        auto it_s_t = Tensor<FP_T>({1, 1, 768}, it_s);

        auto k_h_t = Tensor<FP_T>({1, 16, 5, 48}, k_h);
        auto v_h_t = Tensor<FP_T>({1, 16, 5, 48}, v_h);

        auto ot_s_t = att.forward(it_s_t, &k_h_t, &v_h_t).copy();
  
        compare_data(ot_s_t.data().get(), 
                    ot_s.data(), ot_s.size(), 1e-3);
    }  
    PASSLOG();
}

int main(int argc, char** argv) {
    test_SelfAT();
    return 0;
}

File: cpp/MNN/test/test_selfAT_wh.py
import sys
# pwd 
import os
current_working_directory = os.getcwd()
print("Current working directory:", current_working_directory)

sys.path.append("../python/")  # Adjust the path as necessary

import numpy as np

from MNN import MSelfAT
# set seed for reproducibility
np.random.seed(42)

# embed size 768
at = MSelfAT()


# B S E
input = np.random.uniform(-1, 1, (1, 5, 768)).astype(np.float32)
output, k_h, v_h = at.forward(input)

input_single = np.random.uniform(-1, 1, (1, 1, 768)).astype(np.float32)
output_single, k_h_single, v_h_single = at.forward(input_single, k_h, v_h)

print("Input shape:", input.shape, input.dtype)
print("Output shape:", output.shape, output.dtype)
print("Input strides:", input.strides)
print("Output stride:", output.strides)

save_path = "build/test/temp"
at.k_proj.weight.tofile(save_path + "/k_proj_weight.bin")
at.v_proj.weight.tofile(save_path + "/v_proj_weight.bin")
at.q_proj.weight.tofile(save_path + "/q_proj_weight.bin")
at.out_proj.weight.tofile(save_path + "/out_proj_weight.bin")
at.out_proj.bias.tofile(save_path + "/out_proj_bias.bin")

input_single.tofile(save_path + "/input_single.bin")
output_single.tofile(save_path + "/output_single.bin")
k_h.tofile(save_path + "/k_h.bin")
v_h.tofile(save_path + "/v_h.bin")
k_h_single.tofile(save_path + "/k_h_single.bin")
v_h_single.tofile(save_path + "/v_h_single.bin")
